### PR TITLE
LTD-5012-complex-denial-searching 

### PR DESCRIPTION
**Pull request by depsiatwal**
https://uktrade.atlassian.net/browse/LTD-5012)

This fundamentally changes how we search denials it allows more complex queries which can be customised by the user. 

To validate the default results UAT vs Test Env will use the same data and compare old vs new search/ This will allow us to validate this is sensible to roll out. 

Also dependant on UI PR
https://github.com/uktrade/lite-frontend/pull/1975

### CODE REVIEW

File: api/search/product/views.py [hunk 1]
```diff
@@ -156,6 +156,11 @@ def get_model_queryset(self, document_queryset):
         qs = qs.order_by(preserved_order)
         return qs
 
+    def filter_queryset(self, queryset):
+        qs = super().filter_queryset(queryset)
+        print(qs.to_dict())
```
Comment (kevincarrogan) [inline api/search/product/views.py:161]: There’s debug here.

File: api/external_data/views.py [hunk 2]
```diff
@@ -81,8 +92,32 @@ class DenialSearchView(DocumentViewSet):
 
     def filter_queryset(self, queryset):
         queryset = queryset.filter("term", is_revoked=False).exclude("term", notifying_government="United Kingdom")
+        qs = super().filter_queryset(queryset)
+        print(qs.to_dict())
         return super().filter_queryset(queryset)
 
+    def validate_search_terms(self):
```
Comment (kevincarrogan) [inline api/external_data/views.py:99]: Is this the same as how it’s done on the product view? Could this be extracted out in some way if it’s common.

File: api/core/search/query_backends.py
Comment (markj0hnst0n) [inline api/core/search/query_backends.py:1]: Think this still contains refs to product search not sure if you wanted to make that more generic or if it still applies

File: api/search/product/views.py (hunk 1 above)
Comment (depsiatwal) [reply api/search/product/views.py]: ahh removed

File: api/core/search/query_backends.py
Comment (depsiatwal) [reply api/core/search/query_backends.py:1]: made more generic comment

File: api/external_data/views.py (hunk 2 above)
Comment (depsiatwal) [reply api/external_data/views.py:97]: yeah was kind of thinking the same when I look at it

Comment (depsiatwal) [reply api/external_data/views.py]: shifted to a Mixin

File: api/external_data/views.py [hunk 3]
```diff
@@ -35,25 +37,34 @@ def perform_create(self, serializer):
...
     filter_backends = [
-        filter_backends.SearchFilterBackend,
         filter_backends.SourceBackend,
         filter_backends.FilteringFilterBackend,
+        custom_filter_backends.QueryStringSearchFilterBackend,
         filter_backends.HighlightBackend,
     ]
-    search_fields = ["name", "address", "denial_cle"]
+
+    search_fields = ["name", "address" "item_list_codes"]
```
[the remark this answers is not in the document]
Comment (depsiatwal) [reply api/external_data/views.py]: @kevincarrogan  fixed @markj0hnst0n  it was a merge gone wrong hopefully corrected now

**Review by kevincarrogan — APPROVED** [review]

**Pull request merged**